jrt/src/com/yahoo/jrt/MaybeTlsCryptoSocket.java

```diff
@@ -0,0 +1,126 @@
+        if (data.length != snoop_size) {
+            return false; // wrong data size for tls detection
+        }
+        if (data[0] != 22) {
+            return false; // not tagged as tls handshake
+        }
+        if (data[1] != 3) {
+            return false; // unknown major version
+        }
+        if ((data[2] != 1) && (data[2] != 3)) {
+            return false; // unknown minor version
+        }
+        int frame_len = (data[3] & 0xff);
+        frame_len = ((frame_len << 8) | (data[4] & 0xff));
+        if (frame_len > (16384 + 2048)) {
+            return false; // frame too large
+        }
+        if (frame_len < 4) {
+            return false; // frame too small
+        }
+        if (data[5] != 0x1) {
+            return false; // not tagges as client hello
+        }
+        int hello_len = (data[6] & 0xff);
+        hello_len = ((hello_len << 8) | (data[7] & 0xff));
+        hello_len = ((hello_len << 8) | (data[8] & 0xff));
+        if ((frame_len - 4) != hello_len) {
+            return false; // inconsistent sizes; frame vs client hello
+        }
+        return true;
+    }
+
+    private class MyCryptoSocket extends NullCryptoSocket {
+
+        private TlsCryptoEngine factory;
+        private Buffer buffer;
+
+        MyCryptoSocket(SocketChannel channel, TlsCryptoEngine factory) {
+            super(channel);
+            this.factory = factory;
+            this.buffer = new Buffer(4096);
+        }
+
+        @Override public HandshakeResult handshake() throws IOException {
+            if (factory != null) {
+                channel().read(buffer.getWritable(snoop_size));
+                if (buffer.bytes() < snoop_size) {
+                    return HandshakeResult.NEED_READ;
+                }
+                byte[] data = new byte[snoop_size];
+                ByteBuffer src = buffer.getReadable();
+                for (int i = 0; i < snoop_size; i++) {
+                    data[i] = src.get(i);
+                }
+                if (looksLikeTlsToMe(data)) {
+                    TlsCryptoSocket tlsSocket = factory.createCryptoSocket(channel(), true);
+                    tlsSocket.injectReadData(buffer);
+                    socket = tlsSocket;
+                    return socket.handshake();
+                } else {
+                    factory = null;
+                }
+            }
+            return HandshakeResult.DONE;
+        }
+
+        @Override public int read(ByteBuffer dst) throws IOException {
+            int drainResult = drain(dst);
+            if (drainResult != 0) {
+                return drainResult;
+            }
+            return super.read(dst);
+        }
+
+        @Override public int drain(ByteBuffer dst) throws IOException {
+            int cnt = 0;
+            if (buffer != null) {
+                ByteBuffer src = buffer.getReadable();
+                while (src.hasRemaining() && dst.hasRemaining()) {
```

Comment:
Consider replacing while-loop with `ByteBuffer.put(ByteBuffer src)`

Reply:
that would break the code by throwing an exception if the read buffer (being written to) is too small.

Reply:
Okay, I misunderstood the semantics of `put`.